ffi/src/utils.rs

```diff
@@ -26,8 +26,18 @@ pub unsafe fn w_str_len(s: *const u16) -> usize {
     len
 }
 
-pub unsafe fn raw_str_into_bytes(raw_buffer: *const c_char, len: usize) -> Vec<u8> {
-    unsafe { from_raw_parts(raw_buffer as *const u8, len) }.to_vec()
+/// Converts raw credentials string into [Vec] of bytes.
+///
+/// Credentials are often represented as strings. For example, username, domain, password.
+/// It is OK for Windows SSPI to accept `null` or empty credential strings. The `AcquireCredentialsHandle`
+/// function will return successful status code is we pass the `null` username value. Thus, this function
+/// will return an empty [Vec] in such a case. It is done on purpose to follow the Windows SSPI behaviour.
+pub unsafe fn credentials_str_into_bytes(raw_buffer: *const c_char, len: usize) -> Vec<u8> {
+    if !raw_buffer.is_null() {
+        unsafe { from_raw_parts(raw_buffer as *const u8, len) }.to_vec()
+    } else {
+        Vec::new()
```

Comment:
suggestion: Can you add a safety section in the documentation? In this case, we have the same safety invariants that `from_raw_parts`: https://doc.rust-lang.org/std/vec/struct.Vec.html#safety

I also suggest to properly comment the unsafe blocks explaining how we uphold the invariants.

Examples:
- https://github.com/Devolutions/devolutions-gateway/blob/8ed5163f3a6cf64d789cc43734f10503db88705f/crates/network-scanner/src/lib.rs#L41-L52
- https://github.com/Devolutions/devolutions-gateway/blob/8ed5163f3a6cf64d789cc43734f10503db88705f/crates/win-api-wrappers/src/thread.rs#L33-L36

Of course, updating the code today would be massive work, so let’s just do that incrementally as we modify or add new code.